File: dt-spark-core/src/main/scala/com/kevin/scala/WordFilterDemo.scala
package com.kevin.scala

import org.apache.spark.{SparkConf, SparkContext}

/**
  * 动态统计出现次数最多的单词个数，并过滤掉
  */
object WordFilterDemo {

  def main(args: Array[String]): Unit = {
    val file = "DTSparkCore\\src\\main\\resources\\records.txt"
    // 1.创建SparkConf
    val conf = new SparkConf().setAppName("WordFilterDemo").setMaster("local")
    // 2.创建SparkContext
    val sc = new SparkContext(conf)
    // 3.textFile读取文件数据,flatMap切分数据,map将单词次数初始值为1返回tuple2格式,sample动态抽样数据作为统计
    val sample = sc.textFile(file).flatMap(_.split("\n")).map(line => (line,1)).sample(true,0.5)
    // 4.reduceByKey将相同的单词次数累加,map将次数作为key,sortByKey降序排序,map重新将单词作为key
    val sortMap = sample.reduceByKey(_+_).map(line => (line._2,line._1)).sortByKey(false).map(value => (value._2,value._1))
    // 5.take获取次数最多的一条数据
    val take:Array[(String,Int)] = sortMap.take(1)
    // 打印最多的那条数据
    println(take(0))
    // 6.filter过滤掉最多的一条数据,foreach遍历,println打印
    sortMap.filter(value => !value._1.equals(take(0)._1)).foreach(println(_))
    // 7.关闭sc
    sc.stop()

  }

}

File: dt-spark-core/src/main/java/com/kevin/java/WordFilterDemo.java
package com.kevin.java;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.*;
import scala.Tuple2;

import java.util.Arrays;
import java.util.List;

/**
 * @author kevin
 * @version 1.0
 * @description     动态统计出现次数最多的单词个数，并过滤掉
 * @createDate 2019/1/2
 */
public class WordFilterDemo {

    public static void main(String[] args) {

        SparkConf conf = new SparkConf().setAppName("Demo").setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);
        String file = "DTSparkCore\\src\\main\\resources\\records.txt";
        JavaRDD<String> lines = sc.textFile(file);

        // 切分数据
        JavaRDD<String> flatMap = lines.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public Iterable<String> call(String s) throws Exception {
                return Arrays.asList(s.split("\n"));
            }
        });

        // 对数据初始化值为1
        JavaPairRDD<String, Integer> mapToPair = flatMap.mapToPair(new PairFunction<String, String, Integer>() {
            @Override
            public Tuple2<String, Integer> call(String s) throws Exception {
                return new Tuple2<String, Integer>(s, 1);
            }
        });

        // 动态抽样数据，作为统计
        JavaPairRDD<String, Integer> sample = mapToPair.sample(true, 0.5);

        // 将相同的单词次数累加，并排序
        JavaPairRDD<String, Integer> sortMap = sample.reduceByKey(new Function2<Integer, Integer, Integer>() {
            // 使用reduceByKey算子，将相同的单词次数累加
            @Override
            public Integer call(Integer t1, Integer t2) throws Exception {
                return t1 + t2;
            }
        }).mapToPair(new PairFunction<Tuple2<String, Integer>, Integer, String>() {
            // 使用mapToPair算子，将次数作为key，方便排序
            @Override
            public Tuple2<Integer, String> call(Tuple2<String, Integer> s) throws Exception {
                return new Tuple2<Integer, String>(s._2, s._1);
            }
        }).sortByKey(false).mapToPair(new PairFunction<Tuple2<Integer, String>, String, Integer>() {
            // 降序，将重新将单词做为key
            @Override
            public Tuple2<String, Integer> call(Tuple2<Integer, String> s) throws Exception {
                return new Tuple2<String, Integer>(s._2, s._1);
            }
        });

        // 获取单词次数最多的一条数据
        List<Tuple2<String, Integer>> take = sortMap.take(1);

        System.out.println("动态统计出现次数最多的单词: " + take);

        // 过滤掉次数最多的一条数据
        JavaPairRDD<String, Integer> filter = sortMap.filter(new Function<Tuple2<String, Integer>, Boolean>() {
            // 判断数据是否为最多的那条，如果是，则过滤掉，true保留，false过滤
            @Override
            public Boolean call(Tuple2<String, Integer> s) throws Exception {
                return !s._1.equals(take.get(0)._1);
            }
        });

        // 遍历输出
        filter.foreach(new VoidFunction<Tuple2<String, Integer>>() {
            @Override
            public void call(Tuple2<String, Integer> s) throws Exception {
                System.out.println(s);
            }
        });

        // 关闭
        sc.close();

    }

}
